llvm/lib/CodeGen/GlobalISel/CombinerHelper.cpp

```diff
@@ -6548,6 +6534,52 @@ bool CombinerHelper::tryFoldBoolSelectToLogic(GSelect *Select,
   return false;
 }
 
+bool CombinerHelper::tryFoldSelectOfBinOps(GSelect *Select,
+                                           BuildFnTy &MatchInfo) {
+  Register DstReg = Select->getReg(0);
+  Register Cond = Select->getCondReg();
+  Register False = Select->getFalseReg();
+  Register True = Select->getTrueReg();
+  LLT DstTy = MRI.getType(DstReg);
+
+  GBinOp *LHS = getOpcodeDef<GBinOp>(True, MRI);
+  GBinOp *RHS = getOpcodeDef<GBinOp>(False, MRI);
+
+  // We need two binops of the same kind on the true/false registers.
+  if (!LHS || !RHS || LHS->getOpcode() != RHS->getOpcode())
+    return false;
+
+  // Note that there are no constraints on CondTy.
+  unsigned Flags = LHS->getFlags() & RHS->getFlags();
```

Comment:
This doesn't consider the flags on the select itself

Reply:
The DAG combiner intersects the flags of the binops and ignores the select flags.

Reply:
That's also suboptimal for the DAG combiner. I think it should be (LHS & RHS) | Select. Does InstCombine preserve these?

Reply:
It's also quite likely that code was never properly updated for select gaining flags. It was a relatively recent IR change

Reply:
```cpp
NewBinOp->setFlags(N1->getFlags());
NewBinOp->intersectFlagsWith(N2->getFlags());
```
It ignores the select flags. I updated my code locally.